azure-iot-device/azure/iot/device/provisioning/internal/polling_machine.py

```diff
@@ -0,0 +1,325 @@
+            },
+            {
+                "trigger": "_trig_poll",
+                "source": "waiting_to_poll",
+                "dest": "polling",
+                "after": "_query_operation_status",
+            },
+            {
+                "trigger": "_trig_complete",
+                "source": "registering",
+                "dest": "completed",
+                "after": "_call_complete",
+            },
+            {
+                "trigger": "_trig_complete",
+                "source": "waiting_to_poll",
+                "dest": "completed",
+                "after": "_call_complete",
+            },
+            {
+                "trigger": "_trig_complete",
+                "source": "polling",
+                "dest": "completed",
+                "after": "_call_complete",
+            },
+            {
+                "trigger": "_trig_error",
+                "source": "registering",
+                "dest": "error",
+                "after": "_call_error",
+            },
+            {
+                "trigger": "_trig_error",
+                "source": "waiting_to_poll",
+                "dest": "error",
+                "after": "_call_error",
+            },
+            {
+                "trigger": "_trig_error",
+                "source": "polling",
+                "dest": "error",
+                "after": "_call_error",
+            },
+            {"trigger": "_trig_cancel", "source": "disconnected", "dest": None},
+            {
+                "trigger": "_trig_cancel",
+                "source": "registering",
+                "dest": "cancelling",
+                "after": "clear_and_disconnect",
+            },
+            {
+                "trigger": "_trig_cancel",
+                "source": "waiting_to_poll",
+                "dest": "cancelling",
+                "after": "clear_and_disconnect",
+            },
+            {
+                "trigger": "_trig_cancel",
+                "source": "polling",
+                "dest": "cancelling",
+                "after": "clear_and_disconnect",
+            },
+        ]
+
+        def _on_transition_complete(event_data):
+            if not event_data.transition:
+                dest = "[no transition]"
+            else:
+                dest = event_data.transition.dest
+            logger.info(
+                "Transition complete.  Trigger=%s, Dest=%s, result=%s, error=%s",
+                event_data.event.name,
+                dest,
+                str(event_data.result),
+                str(event_data.error),
+            )
+
+        self._state_machine = Machine(
+            model=self,
+            states=states,
+            transitions=transitions,
+            initial="disconnected",
+            send_event=True,  # Use event_data structures to pass transition arguments
+            finalize_event=_on_transition_complete,
+            queued=True,
+        )
+
+    def register(self, callback=None):
```

Comment:
As an API for this layer, this ought to have formal doc, with the :param: and everything